### PR TITLE
[SIL] Extracted instruction methods from SILCombiner.

### PR DESCRIPTION
**Pull request by nate-chandler**
`SILCombiner` has a number of conveniences for inserting, replacing, and removing instructions that involve modifying a worklist as part of their behavior.  

The conveniences are added to `SILInstructionWorklist` in the first commit ("[SIL] Extracted instruction methods from SILCombiner."). 

In the second commit ("[SILCombiner] Use methods from SILInstructionWorklist."), `SILCombiner` is modified to call through to the methods which were added to `SILInstructionWorklist`.

In the third commit ("[SIL] Tweaked worklist instruction methods' style."), the methods' style is tweaked slightly.

At the moment there is an extra commit stemming from the PR (https://github.com/apple/swift/pull/27028) that this is based on.

### CODE REVIEW

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please ASAN test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

File: include/swift/SIL/SILInstructionWorklist.h [hunk 1]
```diff
@@ -134,86 +134,88 @@ class SILInstructionWorklist : SILInstructionWorklistBase {
...
-    if (Result != I) {
-      assert(&*std::prev(SILBasicBlock::iterator(I)) == Result &&
-            "Expected new instruction inserted before existing instruction!");
-
-#ifndef NDEBUG
-      withStream([&](llvm::raw_ostream &stream, StringRef loggingName) {
-        stream << loggingName << ": Old = " << *I      << '\n'
-               << "  "        << "  New = " << *Result << '\n';
+    if (result != instruction) {
+      assert(&*std::prev(SILBasicBlock::iterator(instruction)) == result &&
```
Comment (gottesmm) [inline include/swift/SIL/SILInstructionWorklist.h]: Just an FYI for future reference (I am fine if you want to ignore this for now), but the *canonical* way to get an iterator for a SILInstruction is to call the method instruction->getIterator().

**Review by gottesmm — APPROVED** [review]

File: include/swift/SIL/SILInstructionWorklist.h (hunk 1 above)
Comment (nate-chandler) [reply include/swift/SIL/SILInstructionWorklist.h]: Noted!

Comment (nate-chandler) [reply include/swift/SIL/SILInstructionWorklist.h]: Fixed.

**Comment by nate-chandler** [comment]
@swift-ci please smoke test and merge

**Comment by nate-chandler** [comment]
@swift-ci please smoke test and merge

**Pull request merged**